pyvips/generate_type_stubs.py

```diff
@@ -0,0 +1,456 @@
+    def write_to_memory(self) -> bytes: ...
+    def write(self, other: Image) -> None: ...
+
+    # Utility methods
+    def invalidate(self) -> None: ...
+    def set_progress(self, progress: bool) -> None: ...
+    def set_kill(self, kill: bool) -> None: ...
+    def copy(self, **kwargs: object) -> Image: ...
+    def tolist(self) -> List[List[float]]: ...
+    # numpy is optional dependency - use TYPE_CHECKING guard
+    def __array__(self, dtype: Optional[str] = None, copy: Optional[bool] = None) -> object: ...
+    def numpy(self, dtype: Optional[str] = None) -> object: ...
+
+    # Dynamically generated operations
+'''
+
+    stub += generate_all_image_operations()
+
+    stub += """
+
+    # Operator overloads
+    def __add__(self, other: Union[Image, float, int]) -> Image: ...
+    def __radd__(self, other: Union[float, int]) -> Image: ...
+    def __sub__(self, other: Union[Image, float, int]) -> Image: ...
+    def __rsub__(self, other: Union[float, int]) -> Image: ...
+    def __mul__(self, other: Union[Image, float, int]) -> Image: ...
+    def __rmul__(self, other: Union[float, int]) -> Image: ...
+    def __truediv__(self, other: Union[Image, float, int]) -> Image: ...
+    def __rtruediv__(self, other: Union[float, int]) -> Image: ...
+    def __floordiv__(self, other: Union[Image, float, int]) -> Image: ...
+    def __rfloordiv__(self, other: Union[float, int]) -> Image: ...
+    def __mod__(self, other: Union[Image, float, int]) -> Image: ...
+    def __pow__(self, other: Union[Image, float, int]) -> Image: ...
+    def __rpow__(self, other: Union[Image, float, int]) -> Image: ...
+    def __abs__(self) -> Image: ...
+    def __neg__(self) -> Image: ...
+    def __pos__(self) -> Image: ...
+    def __invert__(self) -> Image: ...
+    def __lshift__(self, other: Union[Image, float, int]) -> Image: ...
+    def __rshift__(self, other: Union[Image, float, int]) -> Image: ...
+    def __and__(self, other: Union[Image, float, int]) -> Image: ...
+    def __rand__(self, other: Union[Image, float, int]) -> Image: ...
+    def __or__(self, other: Union[Image, float, int]) -> Image: ...
+    def __ror__(self, other: Union[Image, float, int]) -> Image: ...
+    def __xor__(self, other: Union[Image, float, int]) -> Image: ...
+    def __rxor__(self, other: Union[Image, float, int]) -> Image: ...
+    def __eq__(self, other: object) -> bool: ...
+    def __ne__(self, other: object) -> bool: ...
+    def __gt__(self, other: Union[Image, float, int]) -> Image: ...
+    def __ge__(self, other: Union[Image, float, int]) -> Image: ...
+    def __lt__(self, other: Union[Image, float, int]) -> Image: ...
+    def __le__(self, other: Union[Image, float, int]) -> Image: ...
+
+    def __getitem__(self, arg: Union[int, slice, List[int], List[bool]]) -> Image: ...
+    def __call__(self, x: int, y: int) -> List[float]: ...
+    def __repr__(self) -> str: ...
+
+
```

Comment:
There are a set of hand-written bindings too, perhaps some of them could have type hints? 

Easy: `image.floor()` etc.

Medium: `bandjoin`, `bandsplit`, `composite` etc.

Horrible: `ifthenelse`.